openlibrary/plugins/worksearch/code.py

```diff
@@ -202,64 +205,73 @@ def run_solr_query(param = {}, rows=100, page=1, sort=None, spellcheck_count=Non
     if q_list:
         if use_dismax:
-            q = web.urlquote(' '.join(q_list))
-            solr_select += "&defType=dismax&qf=text+title^5+author_name^5&bf=sqrt(edition_count)^10"
+            params.append(('q', ' '.join(q_list)))
+            params.append(('defType', 'dismax'))
+            params.append(('qf', 'text title^5 author_name^5'))
+            params.append(('bf', 'sqrt(edition_count)^10'))
         else:
-            q = web.urlquote(' '.join(q_list + ['_val_:"sqrt(edition_count)"^10']))
-        solr_select += "&q=%s" % q
-    solr_select += '&spellcheck=true&spellcheck.count=%d' % spellcheck_count
-    solr_select += "&facet=true&" + '&'.join("facet.field=" + f for f in facet_fields)
+            params.append(('q', ' '.join(q_list + ['_val_:"sqrt(edition_count)"^10'])))
 
     if 'public_scan' in param:
         v = param.pop('public_scan').lower()
         if v in ('true', 'false'):
             if v == 'false':
                 # also constrain on print disabled since the index may not be in sync
                 param.setdefault('print_disabled', 'false')
-            solr_select += '&fq=public_scan_b:%s' % v
+            params.append(('fq', 'public_scan_b:%s' % v))
 
     if 'print_disabled' in param:
         v = param.pop('print_disabled').lower()
         if v in ('true', 'false'):
-            solr_select += '&fq=%ssubject_key:protected_daisy' % ('-' if v == 'false' else '')
+            minus = '-' if v == 'false' else ''
+            params.append(('fq', '%ssubject_key:protected_daisy' % minus))
 
-    k = 'has_fulltext'
-    if k in param:
-        v = param[k].lower()
+    if 'has_fulltext' in param:
+        v = param['has_fulltext'].lower()
         if v not in ('true', 'false'):
-            del param[k]
-        param[k] == v
-        solr_select += '&fq=%s:%s' % (k, v)
-
-    for k in facet_list_fields:
-        if k == 'author_facet':
-            k = 'author_key'
-        if k not in param:
+            del param['has_fulltext']
+        params.append(('fq', 'has_fulltext:%s' % v))
+
+    for field in facet_list_fields:
+        if field == 'author_facet':
+            field = 'author_key'
+        if field not in param:
             continue
-        v = param[k]
-        solr_select += ''.join('&fq=%s:"%s"' % (k, url_quote(l)) for l in v if l)
-    if sort:
-        solr_select += "&sort=" + url_quote(sort)
+        values = param[field]
+        params += [('fq', '%s:"%s"' % (field, val)) for val in values if val]
 
-    solr_select += '&wt=' + url_quote(param.get('wt', 'standard'))
+    if sort:
+        params.append(('sort', sort))
 
-    solr_result = execute_solr_query(solr_select)
+    params.append(('wt', param.get('wt', 'standard')))
+    params = [(k, v.encode('utf-8') if isinstance(v, six.string_types) else v)
+              for (k, v) in params]
+    url = solr_select_url + '?' + urlencode(params)
+    solr_result = execute_solr_query(url)
```

Comment:
@cclauss Is this the correct py2/py3 way to do this?

Reply:
I don't get what these lines are doing, it looks like it is trying to send url encoded  _binary_ data in the params? 
Can't Solr accept utf-8 parameters?   This seems crazily complex for passing parameters.

Reply:
Or, the issue was with `urlencode`, I believe. We previously did this for the search query here: https://github.com/internetarchive/openlibrary/pull/3117/files#diff-d6b3e6121c27744cb7a743c51608273eL12 , but only for the `q` param. I was getting errors for things like `amélie`, so it looks like it's necessary :/